src/main/java/org/opensearch/agent/tools/SearchAnomalyDetectorsTool.java

```diff
@@ -123,24 +131,92 @@ public <T> void run(Map<String, String> parameters, ActionListener<T> listener)
 
         SearchRequest searchDetectorRequest = new SearchRequest().source(searchSourceBuilder);
 
-        if (running != null || disabled != null || failed != null) {
-            // TODO: add a listener to trigger when the first response is received, to trigger the profile API call
-            // to fetch the detector state, etc.
-            // Will need AD client to onboard the profile API first.
-        }
-
         ActionListener<SearchResponse> searchDetectorListener = ActionListener.<SearchResponse>wrap(response -> {
             StringBuilder sb = new StringBuilder();
-            SearchHit[] hits = response.getHits().getHits();
+            List<SearchHit> hits = Arrays.asList(response.getHits().getHits());
+            Map<String, SearchHit> hitsAsMap = hits.stream().collect(Collectors.toMap(SearchHit::getId, hit -> hit));
+
+            // If we need to filter by detector state, make subsequent profile API calls to each detector
+            if (running != null || disabled != null || failed != null) {
+                List<CompletableFuture<GetAnomalyDetectorResponse>> profileFutures = new ArrayList<>();
+                for (SearchHit hit : hits) {
+                    CompletableFuture<GetAnomalyDetectorResponse> profileFuture = new CompletableFuture<>();
+                    profileFutures.add(profileFuture);
+                    ActionListener<GetAnomalyDetectorResponse> profileListener = ActionListener
+                        .<GetAnomalyDetectorResponse>wrap(profileResponse -> {
+                            profileFuture.complete(profileResponse);
+                        }, e -> {
+                            log.error("Failed to get anomaly detector profile.", e);
+                            profileFuture.completeExceptionally(e);
+                            listener.onFailure(e);
+                        });
+
+                    GetAnomalyDetectorRequest profileRequest = new GetAnomalyDetectorRequest(
+                        hit.getId(),
+                        Versions.MATCH_ANY,
+                        true,
+                        true,
+                        "",
+                        "",
+                        false,
+                        null
+                    );
+                    adClient.getDetectorProfile(profileRequest, profileListener);
```

Comment:
can we move this `getDetectorProfile` call inside a try catch block?

Reply:
This won't throw anything since its a wrapper around `client.execute()` that is void since it is just sending out an async request. The error handling is done for these profile calls is on lines 149-151 if the `profileListener` returns `onFailure()`.